hazelcast/src/main/java/com/hazelcast/internal/util/JVMUtil.java

```diff
@@ -112,6 +113,36 @@ public static long usedMemory(final Runtime runtime) {
         return used;
     }
 
+    /**
+     * Explicit cast to {@link Buffer} parent buffer type. It resolves issues with covariant return types in Java 9+ for
+     * {@link java.nio.ByteBuffer} and {@link java.nio.CharBuffer}. Explicit casting resolves the NoSuchMethodErrors (e.g
+     * java.lang.NoSuchMethodError: java.nio.ByteBuffer.limit(I)Ljava/nio/ByteBuffer) when the project is compiled with newer
+     * Java version and run on Java 8.
+     * <p/>
+     * <a href="https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html">Java 8</a> doesn't provide override the
+     * following Buffer methods in subclasses:
+     *
+     * <pre>
+     * Buffer clear​()
+     * Buffer flip​()
+     * Buffer limit​(int newLimit)
+     * Buffer mark​()
+     * Buffer position​(int newPosition)
+     * Buffer reset​()
+     * Buffer rewind​()
+     * </pre>
+     *
+     * <a href="https://docs.oracle.com/javase/9/docs/api/java/nio/ByteBuffer.html">Java 9</a> introduces the overrides in child
+     * classes (e.g the ByteBuffer), but the return type is the specialized one and not the abstract {@link Buffer}. So the code
+     * compiled with newer Java is not working on Java 8 unless a workaround with explicit casting is used.
+     *
+     * @param buf buffer to cast to the abstract {@link Buffer} parent type
+     * @return the provided buffer
+     */
+    public static Buffer upcast(Buffer buf) {
```

Comment:
Just a question.

I see that, instead of

```java
        ByteBuffer buffer = ByteBuffer.allocate(1024);
        buffer.flip();
``` 

If I use a code like below, 

```java
        Buffer buffer = ByteBuffer.allocate(1024);
        buffer.flip();
```

the generated bytecodes with Java 8 and 11 are the same.
Java 8
```
       0: sipush        1024
       3: invokestatic  #2                  // Method java/nio/ByteBuffer.allocate:(I)Ljava/nio/ByteBuffer;
       6: astore_1
       7: aload_1
       8: invokevirtual #3                  // Method java/nio/Buffer.flip:()Ljava/nio/Buffer;

```

Java 11

```
       0: sipush        1024
       3: invokestatic  #7                  // Method java/nio/ByteBuffer.allocate:(I)Ljava/nio/ByteBuffer;
       6: astore_1
       7: aload_1
       8: invokevirtual #13                 // Method java/nio/Buffer.flip:()Ljava/nio/Buffer;

```

So, I think that another possibility to solve this problem would be to convert all field/variable declarations of type `ByteBuffer` to `Buffer` and eliminate a call to `upcast` every time we want to call the operations you mentioned in the Javadoc. Would that be better?

Reply:
Right, I thought these might be defined in the Buffer class, but that wasn't the case